Dockerfile

```diff
@@ -56,11 +23,89 @@ FROM ${BASE_IMAGE}@${BASE_DIGEST_ARM64} as base-arm64
 ARG BASE_DIGEST_ARM64
 ARG BASE_DIGEST="${BASE_DIGEST_ARM64}"
 
+### Build JRE using the base alpine image
+# hadolint ignore=DL3006
+FROM base-${TARGETARCH} as jre-build
+WORKDIR /
+
+ARG JAVA_VERSION=17.0.7
+
+# Install Zulu directly from the official repository; this could also be the Liberica distribution
+# hadolint ignore=DL3018,DL4006
+RUN --mount=type=cache,target=/var/cache/apk,id=apk,sharing=locked \
+   wget -q -P /etc/apk/keys/ https://cdn.azul.com/public_keys/alpine-signing@azul.com-5d5dc44c.rsa.pub && \
+   echo "https://repos.azul.com/zulu/alpine" | tee -a /etc/apk/repositories && \
+   apk update && apk add binutils "zulu17-jdk-headless~=$JAVA_VERSION"
+
+# Build a custom JRE which will strip down and compress modules to end up with a smaller Java \
+# distribution than the official Zulu JRE. This JRE will also include useful debugging tools like
+# jcmd, jmod, jps, etc., which take little to no space. Anecdotally, compressing modules add around
+# 10ms to the start up time, which is negligible considering our application takes ~10s to start up.
+# See https://adoptium.net/blog/2021/10/jlink-to-produce-own-runtime/
+RUN --mount=type=cache,target=/var/cache/apk,id=apk,sharing=locked \
+   jlink \
+     --add-modules ALL-MODULE-PATH \
+     --strip-debug \
+     --no-man-pages \
+     --no-header-files \
+     --compress=2 \
+     --output /jre && \
+   rm -rf /jre/lib/src.zip
+
+### Java base ###
+# hadolint ignore=DL3006
+FROM base-${TARGETARCH} as java
+WORKDIR /
+
+# Inherit from previous build stage
+ARG JAVA_HOME=/opt/java/openjdk
+
+# Default to UTF-8 file.encoding
+ENV LANG='en_US.UTF-8' LANGUAGE='en_US:en' LC_ALL='en_US.UTF-8'
+
+# Setup JAVA_HOME and binaries in the path
+ENV JAVA_HOME ${JAVA_HOME}
+ENV PATH $JAVA_HOME/bin:$PATH
+
+# Install missing dependencies and update existing ones to fix any pending security issues
+# hadolint ignore=DL3018
+RUN --mount=type=cache,target=/var/cache/apk,id=apk,sharing=locked \
+    ln -s /var/cache/apk /etc/apk/cache && \
+    apk update && apk upgrade && \
+    apk add java-common java-cacerts musl musl-locales musl-locales-lang tzdata zlib
+
+# Copy JRE from previous build stage
+COPY --from=jre-build /jre ${JAVA_HOME}
+
+### Extract zeebe from distball ###
+# hadolint ignore=DL3006
+FROM base-${TARGETARCH} as distball
+WORKDIR /zeebe
+ARG DISTBALL="dist/target/camunda-zeebe-*.tar.gz"
+COPY --link ${DISTBALL} zeebe.tar.gz
+RUN mkdir camunda-zeebe && tar xfvz zeebe.tar.gz --strip 1 -C camunda-zeebe
+
+### Build zeebe from scratch ###
+# We use the latest version here to avoid having to track the Java version multiple times
+# hadolint ignore=DL3006,DL3007
+FROM azul/zulu-openjdk:latest as build
```

Comment:
❌   building locally I get
```
 > [build 4/5] RUN --mount=type=cache,target=/root/.m2,rw ./mvnw -B -am -pl dist package -T1C -D skipChecks -D skipTests:
#17 0.219 Unrecognized option: --add-exports
#17 0.219 Error: Could not create the Java Virtual Machine.
#17 0.219 Error: A fatal exception has occurred. Program will exi
```
using `azul/zulu-openjdk:17-latest` resolved it.

Fun fact the `latest` tag == `8-latest` 🙃 

```suggestion
FROM azul/zulu-openjdk:${JAVA_VERSION} as build
```

Note: this will also require declaring `ARG JAVA_VERSION=17.0.7` globally

On the other hand `bellsoft/liberica-openjdk-alpine:latest` points to the latest 20.0.1 😄

Reply:
Did you try it with the Liberica on Alpine? I think they actually use glibc so maybe it would compile :smile: Didn't with Zulu, which uses musl (since there's no pre-built protoc binary for musl packaged with the library). We can also just use the Java version I suppose :shrug:

Reply:
just tried with `bellsoft/liberica-openjdk-alpine:latest` it works